### DISCUSSION TITLE
The information of Request policy is missing in "Document"

### DISCUSSION THREAD

**Discussion by lyr3051** [Axway-API-Management-Plus/apigateway-openlogging-elk#183] [Q&A] [answered]
I call a API which is register in API Manager and add Request policy "IP White-list".
I found the information of request policy is missing in traffic-summary index:
<img width="949" alt="截圖 2022-03-04 下午4 43 28" src="https://user-images.githubusercontent.com/43023893/156729576-76ffeb88-0fd4-4a4a-a6b1-3d8e156cdddf.png">

I try to find my open traffic log, the Request policy (IP White-list) is still exist.
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/43023893/156728694-6f2ef44b-ae12-4181-bdf4-f3d832f95887.png">

Is the information about Request policy will transfer to elasticsearch by default ? or I should do some configuration to make it work?

**Comment by cwiechmann** [comment]
Hi @lyr3051, The API-Details, incl. what kind of Policies are configured is not directly taken from the OpenTraffic-Event Log. It is retrieved by Logstash during ingestion using the API-Builder REST-API "API-Lookup". 

However, for obvious performance reasons during ingestion into Elasticsearch, API-Details aren't looked up for every single API-Request processed by  Logstash. Instead the information is stored in Memcache and APIBuilder4Elastic for 10 minutes by default.

So, maybe you have called the same API, but without a policy configured, before. With that the API-Details are cached and any further request to the same API in the next 10 minutes, will use the cached API-Details, without your policy.
In other words, wait some time, and you should see the API incl. your policies. There is no additional configuration required.

FYI: What you see in the screenshot is the entire policy-execution path, which is shown in Traffic-Monitor when selecting an API. It's way too complicated to extract anything useful out of it, as it may contain any weird combination of policies.

**Reply by lyr3051** [reply]
Hi @cwiechmann 

I wait some time but the apigw-traffic-summary index still missing the information of Request policy when I call the API successfully  (http status code 200) .

However, when I try to use the wrong path to call the same API, the information of the Request policy finally show up (http status code 403) .

[requestpolicy_index-info.zip](https://github.com/Axway-API-Management-Plus/apigateway-openlogging-elk/files/8194276/requestpolicy_index-info.zip)

**Reply by cwiechmann** [reply]
The `apiPath` is used as the primary cache-key and therefore it might makes sense to see a difference here. However, after some time, the existing information should have been updated anyway.

To validate this, you can restart memcached and APIBuilder4Elastic. This will clean up all caches and you should see the expected result also for the 200 API-Request.

**Reply by lyr3051** [reply]
Hi @cwiechmann 
Thanks.
After I restart memcached and API-Builder, the apiSecurity and request policy of API-details are all display successfully.
Do I have to restart these two components very often for a production environment ? or how long you suggest to restart them?

**Reply by cwiechmann** [reply]
Normally, there is no need to restart anything of the solution. Let me check on side, if I can replicate the behavior, that cached information is not expiring.

**Reply by lyr3051** [reply]
Hi @cwiechmann
Did you replicate this behavior?

I want to confirm what I understand what you are talking about is right or not.
If I call the API with "passthrough" first, but I change passthrough to API Key and call it again within ten minutes, Logstash still retrieves "passthrough" instead of "API key" right?
If yes, that wrong API details "passthrough" will be indexed and stored in elasticsearch, will this security policy of indexed document in be updated to "API key" after 10 minutes? 
Or it will hold the wrong value until it is removed.

**Reply by cwiechmann** [reply]
No, I didn't had the time yet, to confirm it. 

Your understanding is correct. That is the downside of using a cache, but without it, the document ingestion process would be slow down dramatically making it impossible to ingest for instance 1000 API-Requests TPS. 
No, the document which is already indexed wont be updated later, because this is essentially not the way Elasticsearch works. However, as things like API-Security, Policies, etc. are not changed very often and the information is used for Analytics, I believe it is acceptable.

**Reply by Wei-Yo** [reply]
> Hi @lyr3051, The API-Details, incl. what kind of Policies are configured is not directly taken from the OpenTraffic-Event Log. It is retrieved by Logstash during ingestion using the API-Builder REST-API "API-Lookup".
> 
> However, for obvious performance reasons during ingestion into Elasticsearch, API-Details aren't looked up for every single API-Request processed by Logstash. Instead the information is stored in Memcache and APIBuilder4Elastic for 10 minutes by default.
> 
> So, maybe you have called the same API, but without a policy configured, before. With that the API-Details are cached and any further request to the same API in the next 10 minutes, will use the cached API-Details, without your policy. In other words, wait some time, and you should see the API incl. your policies. There is no additional configuration required.
> 
> FYI: What you see in the screenshot is the entire policy-execution path, which is shown in Traffic-Monitor when selecting an API. It's way too complicated to extract anything useful out of it, as it may contain any weird combination of policies.

Hello @cwiechmann 

I have a question for open traffic log,

Why can't the payload of the open traffic log directly display the content of the payload, but point to the file path, and then read the file?
![image](https://user-images.githubusercontent.com/32199735/157809132-1eb843cc-d082-487c-b6ee-11aaf90ccaf7.png)

**Reply by cwiechmann** [reply]
The reason is performance. Payload can be very small, but also very big at some customers. Ingesting payload into Elasticsearch would take very long and reduce the ingestion rate, which makes it impossible to handle 1.000 TPS or more.

**Reply by Wei-Yo** [reply]
oh~ I get it, thanks Chris!
you explained in great detail~

Another question I want to ask is,

Regarding the size of opentrafficlog and the size of payload, is there a file size limit or a limit on the number of files?
Where can I go to adjust this limit?

The relevant settings I can see are only here, are there any other places?
![image](https://user-images.githubusercontent.com/32199735/158331720-311240c3-5fce-4e44-a137-e032735edd48.png)

**Reply by cwiechmann** [reply]
You need to make sure, to configure the OpenTraffic-Eventlog disk space big enough. Read more here: 
https://github.com/Axway-API-Management-Plus/apigateway-openlogging-elk#enable-open-traffic-event-log

The payload is out of band, as you are storing them on a shared volume, which make the payload available to API-Builder4Elastic:
https://github.com/Axway-API-Management-Plus/apigateway-openlogging-elk#traffic-payload
There is no limit for the payload size. 
However, one small difference fyi: 
When using the "Show" payload button in the Traffic-Monitor UI, by default only the first 20.000 bytes are returned, because more data breaks the UX in the UI anyway. If a user really needs to see the entire file, he can use the "Save" option.

**Reply by Wei-Yo** [reply]
I did not find a place to configure the number of OpenTrafficlog and the size and number of a single file of the payload from the link you provided.

Could it be that OpenTraffic log and payload do not rotate when the capacity limit is reached?

**Reply by cwiechmann** [reply]
The size of the OpenTraffic files are static. I think 10MB. You only configure the total disk space for all OpenTraffic files. They are rotated based on the disk space setting you see in the dialog.

Payload files are indeed not rotated. You need to run a script that removes files older than x days.

**Reply by Wei-Yo** [reply]
Thanks for your reply, I see.

In addition, I had some tests for Traffic Monitor earlier, and my configuration is as follows:
![image](https://user-images.githubusercontent.com/32199735/158531282-051f85d6-7201-4fd4-b638-ec6520976eea.png)

When I searched Traffic Monitor for 2021/12/12 Traffic, I didn't find any matching data.
![2](https://user-images.githubusercontent.com/32199735/158532333-53ff55f2-37ee-4e1b-b0bd-e8a167a0ccf6.png)
But I do find 2021/12/12Traffic in the opsdb folder
![1](https://user-images.githubusercontent.com/32199735/158532386-5f273054-3107-439a-8c45-c95b0abe93c3.png)

To sum up the above, I would like to confirm two things.
1. Do you know if the source of Traffic Monitor is opsdb, what is the upper limit of the data that the UI can extract?
2. Will I encounter similar problems if the source of Traffic Monitor is Elasticsearch?

**Comment by lyr3051** [comment]
Hi @cwiechmann 
There is another question about the API-Details.
When I call the API with API Key, the apiSecurity still "Passthrough".
Is the API Security also retrieved by Logstash during ingestion using the API-Builder REST-API "API-Lookup" ?

**Reply by cwiechmann** [reply]
Yes, The API-Details that are looked up also contain the API-Security. It's essentially everything what you configured for API in the API-Manager Frondend-API section.

**Comment by lyr3051** [comment, marked as answer]
Hi @cwiechmann 

I wonder why the request policy is missing in traffic-summary index  before I restart the API-builder and memcached,
but it is still exist in API Gateway Traffic monitoring?


Which index is restrieved from API Builder to show the data in API Gateway Traffic Monitoring?

**Reply by cwiechmann** [reply]
Hi @lyr3051, I believe you discovered a bug in the solution. I realized the same behavior and the root cause, seems the internal In-Memory caching in APIBuilder4Elastic, which never expired.
I fixed this and it will be delivered as part of the next release.
